tests/robustness/validate/validate.go

```diff
@@ -52,6 +56,155 @@ type Config struct {
+					return fmt.Errorf("non unique put %v, required to patch operation history", kv)
+				}
+				putValue[kv] = struct{}{}
+			}
+		}
+	}
+	return nil
+}
+
+func emptyDatabaseAtStart(reports []report.ClientReport) error {
+	for _, r := range reports {
+		for _, op := range r.KeyValue {
+			response := op.Output.(model.MaybeEtcdResponse)
+			if response.Revision == 2 {
+				return nil
+			}
+		}
+	}
+	return fmt.Errorf("non empty database at start or first write didn't succeed, required by model implementation")
+}
+
+func lastOperationSucceededAndObservedInWatch(reports []report.ClientReport) error {
+	var lastOperation porcupine.Operation
+
+	for _, r := range reports {
+		for _, op := range r.KeyValue {
+			if op.Call > lastOperation.Call {
+				lastOperation = op
+			}
+		}
+	}
+	response := lastOperation.Output.(model.MaybeEtcdResponse)
+	if response.PartialResponse || response.Error != "" {
+		return fmt.Errorf("last operation %v failed, required to validate watch", lastOperation)
+	}
+	request := lastOperation.Input.(model.EtcdRequest)
+	if request.Type != model.Txn || len(request.Txn.Conditions) != 0 || len(request.Txn.OperationsOnSuccess) != 1 || request.Txn.OperationsOnSuccess[0].Type != model.PutOperation || request.Txn.OperationsOnSuccess[0].Put.Key != "tombstone" {
+		return fmt.Errorf("last operation %+v didn't write `end` key, required to validate watch", request)
+	}
+	for _, r := range reports {
+		for _, watch := range r.Watch {
+			for _, watchResp := range watch.Responses {
+				for _, e := range watchResp.Events {
+					if e.Revision > response.Revision {
+						return fmt.Errorf("observed watch event %v after last operation, required to validate watch", e)
+					}
+					if e.Revision == response.Revision && e.Key != "tombstone" {
+						return fmt.Errorf("observed watch event %v different than last operation, required to validate watch", e)
+					}
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func observedAllRevisionsInWatch(reports []report.ClientReport) error {
+	var maxRevision int64
+	for _, r := range reports {
+		for _, watch := range r.Watch {
+			for _, watchResp := range watch.Responses {
+				for _, e := range watchResp.Events {
+					if e.Revision > maxRevision {
+						maxRevision = e.Revision
+					}
+				}
+			}
+		}
+	}
+	observedRevisions := make([]bool, maxRevision+1)
+	for _, r := range reports {
+		for _, watch := range r.Watch {
+			for _, watchResp := range watch.Responses {
+				for _, e := range watchResp.Events {
+					observedRevisions[e.Revision] = true
+				}
+			}
+		}
+	}
+	for i := 2; i < len(observedRevisions); i++ {
+		if !observedRevisions[i] {
+			return fmt.Errorf("didn't observe revision %d in watch, required to patch operation and validate serializable requests", i)
+		}
+	}
+	return nil
+}
+
+func nonConcurrentClientRequests(reports []report.ClientReport) error {
+	lastClientRequestReturn := map[int]int64{}
+	for _, r := range reports {
+		for _, op := range r.KeyValue {
```

Comment:
question: could 2 different ClientReport contain ops from the same client? 
If not, why is the map needed? In that case, you just need to set lastRequest=0 at the beginning of each report. 
If yes, the time ordering between different reports are not guaranteed.

Reply:
They should not, but it's possible if the report got corrupted. The goal is to detect such things early as linearization will return broken hard to decode results. 

> If yes, the time ordering between different reports are not guaranteed. 

Right, but it's ok to assume so.

Reply:
Thanks for the clarification. Can you add some comments regarding this? It might not be clear to future contributors digging into the code why the map is needed.